text/0000-prepare-operation.md

```diff
@@ -0,0 +1,374 @@
+
+1. [Moving `io.buildpacks` properties to `io.buildpacks.defaults`.](#Namespace-iobuildpacksdefaults)
+2. [A replaceable new phase `prepare` that applies platform configuration.](#Preparer)
+3. [Supporting Cloud Native Buildpack utilities.](#Cloud-Native-Buildpacks-utilities)
+
+## Changes to `io.buildpacks` namespace
+
+### Namespace `io.buildpacks.defaults`
+
+This namespace declares default values that platforms should acknowledge. A majority of properties should be spec'd but we would allow for unspec'd additional properties.
+
+The criteria for spec'd properties in this namespace would be based on whether they affect inputs to the lifecycle phases. For example, the order and environment variables are inputs to the lifecycle. The `builder`, in contrast, is not an input to the lifecycle so it wouldn't meet the criteria.
+
+Therefore the following changes have been made from the [latest schema][pd-02]:
+
+ - Remove `builder`
+
+[pd-02]: https://github.com/buildpacks/spec/blob/extensions/project-descriptor/0.2/extensions/project-descriptor.md#iobuildpacks-optional
+
+### Version `io.buildpacks` namespace
+
+Given that the project descriptor's `io.buildpacks` namespace is directly tied to the platform by the fact that it's the platform that needs to consume and parse it, it makes sense to unify the version of the schema with that of the platform API.
+
+
+### Schema
+
+See [Platform Spec changes](#Platform-Spec).
+
+#### Usages Examples
+
+###### Multi-Platform (pack + kpack)
+
+In the following example we show how various platform namespaces can have their own configuration but properties can be promoted to `io.buildpacks.defaults` if the user's intent is to apply it to all platforms.
+
+```toml
+[io.buildpacks]
+schema-version = "0.2"
+
+###
+# common buildpacks config
+##
+
+[io.buildpacks.defaults]
+exclude = ["some-files/**"]
+
+##
+# pack config
+##
+
+[io.buildpacks.pack]
+schema-version = "0.1"
+builder = "cnbs/sample-builder:bionic"
+image = "my-app"                        # (example only)
+run-image = "my-run-image:latest"       # (example only)
+
+##
+# kpack config
+##
+
+[com.vmware.kpack]
+schema-version = "0.1"
+
+[[com.vmware.kpack.build.env]]
+name="CUSTOM_ENV"
+value="SOME_VALUE"
+```
+
+Notice:
+
+1. Both `pack` and `kpack` are expected to apply `io.buildpacks.defaults.exclude`.
+2. `io.buildpacks.pack` has additional properties that are specific to `pack` only.
+
+## Prepare phase
+
+A `prepare` phase would be a new lifecycle phase that gets executed before `create` or `detect` phase. Similar to the `rebaser`, it is outside of the standard [Build operation][build-operation].
+
+[build-operation]: https://github.com/buildpacks/spec/blob/main/platform.md#build
+
+#### Responsibilities
+
+At minimum, the expected reponsibility of the `preparer` would be to:
+
+ - Apply the requested configuration.
+
+#### Inputs
+
+The `prepare` phase would take all the same inputs as `creator` plus the addition of a path to `project.toml`.
```

Comment:
Is this binary intended to run on builder images?

Reply:
Possible, yes. Required no.

Reply:
But it's running either in the builder image or not at all, right?

Potentially related question:would it be called as part of the `creator` lifecycle?

Reply:
> But it's running either in the builder image or not at all, right?

I can't think of a reason why it _must_ run on the builder image. As an example, Tekton already does some preparing step [here](https://github.com/buildpacks/tekton-integration/blob/771bbd2ecec9b671889db48214f4e65eb3e9113e/task/buildpacks-phases/0.2/buildpacks-phases.yaml#L76-L127) using a completely different image. I can foresee platforms wanting to provide their own container for the execution of their preparation mechanism.

Hypothetically, given that ultimately preparing only impacts the build using the existing file system interface, platforms can apply changes "natively" by making changes to the "workspace" on the fly or to volumes that would later be mounted to the next phases.

> Would it be called as part of the creator lifecycle?

No. There's a light mention of that [here](https://github.com/buildpacks/rfcs/pull/202/files/99fc145c93f07e045e62b9c5d42405ecffc53afc#diff-da0c0db1eef841f956d231dff1f0ad819daf2ee72f55b71032073d431b767277R119). This phase is considered to be outside of the Build operation.